pkg/graph/crd/compat/schema.go

```diff
@@ -0,0 +1,220 @@
+			PatternChanged,
+			oldSchema.Pattern,
+			newSchema.Pattern,
+		)
+	}
+
+	// Compare properties
+	compareProperties(path, oldSchema, newSchema, result)
+
+	// Check required fields
+	compareRequiredFields(path, oldSchema, newSchema, result)
+
+	// Check enum values
+	compareEnumValues(path, oldSchema, newSchema, result)
+
+	// For arrays, check items schema
+	compareArrayItems(path, oldSchema, newSchema, result)
+
+	return result, nil
+}
+
+// compareProperties checks for added, removed, or changed properties
+func compareProperties(path string, oldSchema, newSchema *v1.JSONSchemaProps, result *DiffResult) {
+	// First, check for removed properties (breaking changes)
+	for propName, oldProp := range oldSchema.Properties {
+		propPath := path + ".properties." + propName
+
+		// check if property still exists
+		newProp, exists := newSchema.Properties[propName]
+		if !exists {
+			// property was removed - breaking change
+			result.AddBreakingChange(propPath, PropertyRemoved, "", "")
+			continue
+		}
+
+		// property exists in both schemas - compare them recursively
+		propResult, _ := compareSchemas(propPath, &oldProp, &newProp)
+		result.BreakingChanges = append(result.BreakingChanges, propResult.BreakingChanges...)
+		result.NonBreakingChanges = append(result.NonBreakingChanges, propResult.NonBreakingChanges...)
+	}
+
+	// Then check for added properties. Now things get a bit more spicy.
+	// A new property can be required or optional, and can have a default value.
+	// Depending on these factors, it can be a breaking or non-breaking change.
+	//
+	// In general the rules are:
+	// - Adding a required property without a default value is a breaking change
+	// - Adding a required property with a default value is a non-breaking change
+	// - Adding an optional property is a non-breaking change
+
+	newRequiredSet := toStringSet(newSchema.Required)
+
+	for propName, newProp := range newSchema.Properties {
+		if _, exists := oldSchema.Properties[propName]; !exists {
+			propPath := path + ".properties." + propName
+
+			// check if property is required but has a default (non-breaking)
+			// or required without default (breaking)
+			hasDefault := newProp.Default != nil && len(newProp.Default.Raw) > 0
+
+			if newRequiredSet[propName] && !hasDefault {
+				// property is required and has no default - breaking change
+				result.AddBreakingChange(propPath, PropertyAdded, "required=false", "required=true")
+			} else {
+				// property is optional or has default - non-breaking change
+				result.AddNonBreakingChange(propPath, PropertyAdded, "", "")
+			}
+		}
+	}
+}
+
+// compareRequiredFields checks for changes to required fields, it only considers
+// existing properties, since new properties are handled in compareProperties.
+func compareRequiredFields(path string, oldSchema, newSchema *v1.JSONSchemaProps, result *DiffResult) {
+	// Use length checks instead of nil checks
+	if len(oldSchema.Required) == 0 && len(newSchema.Required) == 0 {
```

Comment:
you might want to test that old and new schema are not nil as well just to protect yourself. It is likely someone attempts to delete the schema to work around something.